src/transformStateWithClones.js

```diff
@@ -6,8 +6,50 @@
  *
  * @return {Object[]}
  */
-function transformStateWithClones(state, actions) {
-  // write code here
+function transformStateWithClones(initialState, actions) {
+  if (!Array.isArray(actions)) {
+    throw new Error('Параметр "actions" должен быть массивом.');
+  }
+
+  let currentState = { ...initialState };
+  const states = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+
+      case 'addProperties':
+        if (!action.extraData || typeof action.extraData !== 'object') {
+          throw new Error(
+            'Для действия "addProperties" требуется объект "extraData".',
+          );
+        }
+        currentState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        if (!Array.isArray(action.keysToRemove)) {
+          throw new Error(
+            'Для действия "removeProperties" требуется массив "keysToRemove".',
+          );
+        }
+        currentState = { ...currentState };
+
+        for (const key of action.keysToRemove) {
+          delete currentState[key];
+        }
+        break;
+
+      default:
+        throw new Error(`Неизвестный тип действия: ${action.type}`);
+    }
+
+    states.push({ ...currentState });
```

Comment:
The `states.push({ ...currentState });` line should be moved outside of the `switch` block, after the `switch` statement, to ensure that the state is captured correctly after each action is processed. This aligns with the requirement to push a copy of the object to the array at the end of each loop cycle, but outside of the `switch` block.